tests/spawn-utils.c

```diff
@@ -0,0 +1,345 @@
+
+  return g_spawn_sync (NULL,
+                       argv,
+                       merged_env,
+                       flags,
+                       NULL,
+                       NULL,
+                       out,
+                       err,
+                       status,
+                       error);
+}
+
+gboolean
+test_spawn_cwd (const gchar *cwd,
+                gchar **argv,
+                gchar **envp,
+                CmdStuff *cmd,
+                GError **error)
+{
+  return test_spawn_cwd_full (cwd,
+                              argv,
+                              envp,
+                              cmd == NULL,
+                              cmd,
+                              error);
+}
+
+gboolean
+test_spawn (gchar **argv,
+            gchar **envp,
+            CmdStuff *cmd,
+            GError **error)
+{
+  return test_spawn_cwd (NULL,
+                         argv,
+                         envp,
+                         cmd,
+                         error);
+}
+
+static void
+env_to_hash_table (gchar **envp,
+                   GHashTable *hash_table,
+                   const gchar *desc)
+{
+  gchar **iter;
+
+  for (iter = envp; *iter != NULL; ++iter)
+    {
+      g_auto(GStrv) key_and_value = g_strsplit (*iter, "=", 2);
+
+      if (g_strv_length (key_and_value) != 2)
+        g_error ("Invalid %s environment value %s", desc, *iter);
+      g_hash_table_insert (hash_table,
+                           g_strdup (key_and_value[0]),
+                           g_strdup (key_and_value[1]));
+    }
+}
+
+static gchar **
+hash_table_to_env (GHashTable *hash_table)
+{
+  g_autoptr(GPtrArray) envp = NULL;
+  GHashTableIter hiter;
+  gpointer key_ptr;
+  gpointer value_ptr;
+
+  envp = string_array_new ();
+  g_hash_table_iter_init (&hiter, hash_table);
+  while (g_hash_table_iter_next (&hiter, &key_ptr, &value_ptr))
+    g_ptr_array_add (envp, envvar(key_ptr, value_ptr));
+  g_ptr_array_add (envp, NULL);
+
+  return (gchar **)g_ptr_array_free (g_steal_pointer (&envp), FALSE);
+}
+
+gchar **
+merge_parent_and_child_env (gchar **child_env)
+{
+  g_auto(GStrv) parent = g_get_environ ();
+  g_autoptr(GHashTable) henv = NULL;
+
+  if (child_env == NULL)
+    return g_steal_pointer (&parent);
+
+  henv = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+  env_to_hash_table (parent, henv, "parent");
+  env_to_hash_table (child_env, henv, "child");
+
+  return hash_table_to_env (henv);
+}
+
+typedef struct
+{
+  CmdStuff *cmd;
+  CmdAsyncStuff *async_cmd;
+  GMainLoop *loop;
+  GError **error;
+} ReapData;
+
+static void
+collect_output (GPid pid,
+		gint status,
+		gpointer reap_data_ptr)
+{
+  ReapData *reap_data = reap_data_ptr;
+
+  g_main_loop_quit (reap_data->loop);
+  reap_data->cmd->exit_status = status;
+  if (!input_stream_to_string (reap_data->async_cmd->out_stream,
+                               &reap_data->cmd->standard_output,
+                               reap_data->error))
+    return;
```

Comment:
`collect_output` is a callback for `g_child_watch_add` so error checking is the only option here.